File: local/wikipedia_downloader.py
#!/usr/bin/env python3
"""
Wikipedia Download and Processing System
Downloads Wikipedia dumps and processes them for offline AI integration
"""

import os
import sys
import json
import sqlite3
import requests
import xml.etree.ElementTree as ET
from urllib.parse import urljoin
import bz2
import gzip
import re
import argparse
from datetime import datetime
import logging
from pathlib import Path

# Setup logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

class WikipediaDownloader:
    """Download and process Wikipedia dumps for offline use"""
    
    def __init__(self, data_dir="./wikipedia_data"):
        self.data_dir = Path(data_dir)
        self.data_dir.mkdir(exist_ok=True)
        
        # Wikipedia dump URLs
        self.base_url = "https://dumps.wikimedia.org"
        
        # Available datasets
        self.datasets = {
            'simple': {
                'name': 'Simple English Wikipedia',
                'url': 'simplewiki/latest/simplewiki-latest-pages-articles.xml.bz2',
                'size_mb': 200,
                'articles': 200000,
                'description': 'Simplified English articles, perfect for basic knowledge'
            },
            'featured': {
                'name': 'English Wikipedia Featured Articles',
                'url': 'enwiki/latest/enwiki-latest-pages-articles1.xml-p1p41242.bz2',
                'size_mb': 50,
                'articles': 6000,
                'description': 'High-quality featured articles only'
            },
            'full': {
                'name': 'Full English Wikipedia',
                'url': 'enwiki/latest/enwiki-latest-pages-articles.xml.bz2',
                'size_mb': 20000,
                'articles': 6000000,
                'description': 'Complete English Wikipedia (very large)'
            }
        }
    
    def get_available_datasets(self):
        """Get list of available Wikipedia datasets"""
        return self.datasets
    
    def download_dataset(self, dataset_name, progress_callback=None):
        """Download a Wikipedia dataset"""
        if dataset_name not in self.datasets:
            raise ValueError(f"Unknown dataset: {dataset_name}")
        
        dataset = self.datasets[dataset_name]
        url = urljoin(self.base_url, dataset['url'])
        filename = self.data_dir / f"{dataset_name}_wikipedia.xml.bz2"
        
        logger.info(f"Downloading {dataset['name']} from {url}")
        
        try:
            response = requests.get(url, stream=True)
            response.raise_for_status()
            
            total_size = int(response.headers.get('content-length', 0))
            downloaded = 0
            
            with open(filename, 'wb') as f:
                for chunk in response.iter_content(chunk_size=8192):
                    if chunk:
                        f.write(chunk)
                        downloaded += len(chunk)
                        
                        if progress_callback and total_size > 0:
                            progress = (downloaded / total_size) * 100
                            progress_callback(progress, downloaded, total_size)
            
            logger.info(f"Downloaded {dataset['name']} to {filename}")
            return str(filename)
            
        except Exception as e:
            logger.error(f"Failed to download {dataset['name']}: {e}")
            if filename.exists():
                filename.unlink()
            raise
    
    def extract_and_process(self, compressed_file, db_path, progress_callback=None):
        """Extract and process Wikipedia XML dump into SQLite database"""
        logger.info(f"Processing {compressed_file} into {db_path}")
        
        # Initialize database
        db = WikipediaDatabase(db_path)
        db.initialize()
        
        # Process XML file
        try:
            with bz2.open(compressed_file, 'rt', encoding='utf-8') as f:
                processor = WikipediaXMLProcessor(db, progress_callback)
                processor.process_xml_stream(f)
            
            # Create search indexes
            db.create_search_index()
            
            logger.info("Wikipedia processing completed successfully")
            return db_path
            
        except Exception as e:
            logger.error(f"Failed to process Wikipedia dump: {e}")
            raise

class WikipediaXMLProcessor:
    """Process Wikipedia XML dumps"""
    
    def __init__(self, database, progress_callback=None):
        self.db = database
        self.progress_callback = progress_callback
        self.articles_processed = 0
        
        # Regex patterns for cleaning
        self.cleanup_patterns = [
            (r'\{\{[^}]*\}\}', ''),  # Remove templates
            (r'\[\[Category:[^\]]*\]\]', ''),  # Remove category links
            (r'\[\[File:[^\]]*\]\]', ''),  # Remove file links
            (r'\[\[Image:[^\]]*\]\]', ''),  # Remove image links
            (r'<ref[^>]*>.*?</ref>', ''),  # Remove references
            (r'<ref[^>]*\/>', ''),  # Remove self-closing refs
            (r'&lt;.*?&gt;', ''),  # Remove HTML entities
            (r'\n\s*\n', '\n'),  # Remove extra newlines
        ]
    
    def process_xml_stream(self, xml_file):
        """Process Wikipedia XML stream"""
        logger.info("Starting XML processing...")
        
        # Parse XML incrementally
        context = ET.iterparse(xml_file, events=('start', 'end'))
        context = iter(context)
        event, root = next(context)
        
        current_page = {}
        current_element = None
        
        for event, elem in context:
            current_element = elem.tag.split('}')[-1]  # Get tag name without namespace
            if event == 'end':
                #logger.info(current_element)
                if current_element == 'page':
                    # Process complete page
                    if self.is_valid_article(current_page):
                        self.process_article(current_page)
                        self.articles_processed += 1
                        
                        if self.progress_callback and self.articles_processed % 1000 == 0:
                            self.progress_callback(self.articles_processed)
                    
                    current_page = {}
                    root.clear()  # Free memory
                    
                elif current_element in ['title', 'text', 'id']:
                    current_page[current_element] = elem.text or ''

                current_element = None
    
    def is_valid_article(self, page):
        """Check if page is a valid article"""
        title = page.get('title', '')
        text = page.get('text', '')
        
        # Skip redirects, disambiguation, and special pages
        if (title.startswith('Category:') or 
            title.startswith('Template:') or 
            title.startswith('File:') or
            title.startswith('Wikipedia:') or
            '#REDIRECT' in text.upper() or
            len(text.strip()) < 100):
            return False
        
        return True
    
    def process_article(self, page):
        """Process and store a single article"""
        title = page.get('title', '').strip()
        content = page.get('text', '')
        article_id = page.get('id', '')
        
        # Clean content
        cleaned_content = self.clean_wikipedia_markup(content)
        
        # Extract summary (first paragraph)
        summary = self.extract_summary(cleaned_content)
        
        # Extract categories
        categories = self.extract_categories(content)
        
        # Store in database
        self.db.insert_article(
            article_id=article_id,
            title=title,
            content=cleaned_content,
            summary=summary,
            categories=categories
        )
    
    def clean_wikipedia_markup(self, text):
        """Clean Wikipedia markup from text"""
        # Apply cleanup patterns
        for pattern, replacement in self.cleanup_patterns:
            text = re.sub(pattern, replacement, text, flags=re.DOTALL | re.IGNORECASE)
        
        # Clean wiki links [[Link|Text]] -> Text
        text = re.sub(r'\[\[([^|\]]+\|)?([^\]]+)\]\]', r'\2', text)
        
        # Clean simple formatting
        text = re.sub(r"'''([^']+)'''", r'\1', text)  # Bold
        text = re.sub(r"''([^']+)''", r'\1', text)    # Italic
        
        # Clean up whitespace
        text = re.sub(r'\n\s*\n', '\n\n', text)
        text = text.strip()
        
        return text
    
    def extract_summary(self, content):
        """Extract article summary (first paragraph)"""
        paragraphs = content.split('\n\n')
        for para in paragraphs:
            para = para.strip()
            if len(para) > 50 and not para.startswith('='):
                return para[:500] + ('...' if len(para) > 500 else '')
        return ''
    
    def extract_categories(self, content):
        """Extract categories from article"""
        categories = re.findall(r'\[\[Category:([^\]]+)\]\]', content)
        return json.dumps(categories)

class WikipediaDatabase:
    """Database management for Wikipedia content"""
    
    def __init__(self, db_path):
        self.db_path = db_path
        self.conn = None
    
    def initialize(self):
        """Initialize database schema"""
        self.conn = sqlite3.connect(self.db_path)
        self.conn.execute('PRAGMA journal_mode=WAL')
        self.conn.execute('PRAGMA synchronous=NORMAL')
        self.conn.execute('PRAGMA cache_size=10000')
        
        # Create tables
        self.conn.executescript('''
            CREATE TABLE IF NOT EXISTS wikipedia_articles (
                id INTEGER PRIMARY KEY,
                article_id TEXT UNIQUE,
                title TEXT NOT NULL,
                content TEXT NOT NULL,
                summary TEXT,
                categories TEXT,
                word_count INTEGER,
                created_at DATETIME DEFAULT CURRENT_TIMESTAMP
            );
            
            CREATE TABLE IF NOT EXISTS wikipedia_metadata (
                key TEXT PRIMARY KEY,
                value TEXT
            );
            
            CREATE INDEX IF NOT EXISTS idx_title ON wikipedia_articles(title);
            CREATE INDEX IF NOT EXISTS idx_article_id ON wikipedia_articles(article_id);
        ''')
        
        self.conn.commit()
    
    def insert_article(self, article_id, title, content, summary, categories):
        """Insert article into database"""
        word_count = len(content.split())
        
        self.conn.execute('''
            INSERT OR REPLACE INTO wikipedia_articles 
            (article_id, title, content, summary, categories, word_count)
            VALUES (?, ?, ?, ?, ?, ?)
        ''', (article_id, title, content, summary, categories, word_count))
        
        # Commit every 1000 articles
        if self.conn.total_changes % 1000 == 0:
            self.conn.commit()
    
    def create_search_index(self):
        """Create full-text search index"""
        logger.info("Creating full-text search index...")
        
        self.conn.executescript('''
            CREATE VIRTUAL TABLE IF NOT EXISTS wikipedia_fts USING fts5(
                title, content, summary,
                content='wikipedia_articles',
                content_rowid='id'
            );
            
            INSERT INTO wikipedia_fts(wikipedia_fts) VALUES('rebuild');
        ''')
        
        self.conn.commit()
        logger.info("Search index created successfully")
    
    def get_stats(self):
        """Get database statistics"""
        cursor = self.conn.execute('''
            SELECT 
                COUNT(*) as total_articles,
                SUM(word_count) as total_words,
                AVG(word_count) as avg_words_per_article
            FROM wikipedia_articles
        ''')
        
        return cursor.fetchone()
    
    def close(self):
        """Close database connection"""
        if self.conn:
            self.conn.commit()
            self.conn.close()

def main():
    """Main CLI interface"""
    parser = argparse.ArgumentParser(description='Wikipedia Download and Processing Tool')
    parser.add_argument('--action', choices=['list', 'download', 'process'], required=True,
                       help='Action to perform')
    parser.add_argument('--dataset', choices=['simple', 'featured', 'full'],
                       help='Dataset to download/process')
    parser.add_argument('--data-dir', default='./wikipedia_data',
                       help='Directory for Wikipedia data')
    parser.add_argument('--db-path', default='./wikipedia.db',
                       help='SQLite database path')
    
    args = parser.parse_args()
    
    downloader = WikipediaDownloader(args.data_dir)
    
    if args.action == 'list':
        print("Available Wikipedia datasets:")
        for name, info in downloader.get_available_datasets().items():
            print(f"\n{name}:")
            print(f"  Name: {info['name']}")
            print(f"  Size: ~{info['size_mb']}MB")
            print(f"  Articles: ~{info['articles']:,}")
            print(f"  Description: {info['description']}")
    
    elif args.action == 'download':
        if not args.dataset:
            print("Error: --dataset required for download")
            sys.exit(1)
        
        def progress_callback(progress, downloaded, total):
            print(f"\rDownload progress: {progress:.1f}% ({downloaded:,}/{total:,} bytes)", end='')
        
        try:
            filename = downloader.download_dataset(args.dataset, progress_callback)
            print(f"\nDownload completed: {filename}")
        except Exception as e:
            print(f"\nDownload failed: {e}")
            sys.exit(1)
    
    elif args.action == 'process':
        if not args.dataset:
            print("Error: --dataset required for processing")
            sys.exit(1)
        
        compressed_file = Path(args.data_dir) / f"{args.dataset}_wikipedia.xml.bz2"
        if not compressed_file.exists():
            print(f"Error: {compressed_file} not found. Download first.")
            sys.exit(1)
        
        def progress_callback(articles_processed):
            print(f"\rProcessed {articles_processed:,} articles", end='')
        
        try:
            db_path = downloader.extract_and_process(compressed_file, args.db_path, progress_callback)
            print(f"\nProcessing completed: {db_path}")
            
            # Show statistics
            db = WikipediaDatabase(db_path)
            db.initialize()
            stats = db.get_stats()
            print(f"Stats: {stats}")
            # print(f"Total articles: {stats[0]:,}")
            # print(f"Total words: {stats[1]:,}")
            # print(f"Average words per article: {stats[2]:.1f}")
            db.close()
            
        except Exception as e:
            print(f"\nProcessing failed: {e}")
            sys.exit(1)

if __name__ == '__main__':
    main()

File: core/wikipedia-downloader.js
/**
 * Wikipedia Download and Processing System (Node.js)
 * Downloads Wikipedia dumps and processes them for offline AI integration
 * 
 * Optimized for low-memory environments (Heroku 512MB dynos)
 */

const https = require('https');
const http = require('http');
const fs = require('fs');
const path = require('path');
const unbzip2 = require('unbzip2-stream');
const sax = require('sax');
const sqlite3 = require('sqlite3').verbose();

/**
 * Available Wikipedia datasets
 */
const DATASETS = {
    simple: {
        name: 'Simple English Wikipedia',
        url: 'https://dumps.wikimedia.org/simplewiki/latest/simplewiki-latest-pages-articles.xml.bz2',
        size_mb: 200,
        articles: 200000,
        description: 'Simplified English articles, perfect for basic knowledge'
    },
    featured: {
        name: 'English Wikipedia Featured Articles',
        url: 'https://dumps.wikimedia.org/enwiki/latest/enwiki-latest-pages-articles1.xml-p1p41242.bz2',
        size_mb: 50,
        articles: 6000,
        description: 'High-quality featured articles only'
    },
    full: {
        name: 'Full English Wikipedia',
        url: 'https://dumps.wikimedia.org/enwiki/latest/enwiki-latest-pages-articles.xml.bz2',
        size_mb: 20000,
        articles: 6000000,
        description: 'Complete English Wikipedia (very large)'
    }
};

/**
 * Regex patterns for cleaning Wikipedia markup
 * Pre-compiled for performance
 */
const CLEANUP_PATTERNS = [
    [/\{\{[^}]*\}\}/g, ''],           // Remove templates
    [/\[\[Category:[^\]]*\]\]/gi, ''], // Remove category links
    [/\[\[File:[^\]]*\]\]/gi, ''],     // Remove file links
    [/\[\[Image:[^\]]*\]\]/gi, ''],    // Remove image links
    [/<ref[^>]*>.*?<\/ref>/gis, ''],   // Remove references
    [/<ref[^>]*\/>/gi, ''],            // Remove self-closing refs
    [/&lt;.*?&gt;/g, ''],              // Remove HTML entities
];

/**
 * Memory-efficient text cleaning
 * Processes text in place without creating many intermediate strings
 */
function cleanWikipediaMarkup(text) {
    if (!text || text.length === 0) return '';
    
    // Limit text size to prevent memory issues with very long articles
    const MAX_CONTENT_LENGTH = 50000; // 50KB max per article
    if (text.length > MAX_CONTENT_LENGTH) {
        text = text.substring(0, MAX_CONTENT_LENGTH);
    }
    
    // Apply cleanup patterns
    for (const [pattern, replacement] of CLEANUP_PATTERNS) {
        text = text.replace(pattern, replacement);
    }
    
    // Clean wiki links [[Link|Text]] -> Text
    text = text.replace(/\[\[([^|\]]+\|)?([^\]]+)\]\]/g, '$2');
    
    // Clean simple formatting
    text = text.replace(/'''([^']+)'''/g, '$1');  // Bold
    text = text.replace(/''([^']+)''/g, '$1');    // Italic
    
    // Clean up whitespace (single pass)
    text = text.replace(/\n\s*\n+/g, '\n\n').trim();
    
    return text;
}

/**
 * Extract summary efficiently
 */
function extractSummary(content) {
    if (!content) return '';
    
    // Find first substantial paragraph without splitting entire content
    let start = 0;
    let end = 0;
    
    while (start < content.length && start < 5000) { // Only search first 5KB
        // Find end of paragraph
        end = content.indexOf('\n\n', start);
        if (end === -1) end = Math.min(content.length, start + 500);
        
        const para = content.substring(start, end).trim();
        
        // Check if this is a valid paragraph
        if (para.length > 50 && !para.startsWith('=')) {
            return para.length > 500 ? para.substring(0, 500) + '...' : para;
        }
        
        start = end + 2;
    }
    
    return '';
}

/**
 * Extract categories efficiently
 */
function extractCategories(content) {
    if (!content) return '[]';
    
    const categories = [];
    const regex = /\[\[Category:([^\]|]+)/gi;
    let match;
    
    // Limit to first 20 categories
    while ((match = regex.exec(content)) !== null && categories.length < 20) {
        categories.push(match[1].trim());
    }
    
    return JSON.stringify(categories);
}

/**
 * WikipediaDownloader class - handles downloading and processing Wikipedia dumps
 */
class WikipediaDownloader {
    constructor(dataDir = './wikipedia_data') {
        this.dataDir = dataDir;
        this.datasets = DATASETS;
        
        // Ensure data directory exists
        if (!fs.existsSync(this.dataDir)) {
            fs.mkdirSync(this.dataDir, { recursive: true });
        }
    }
    
    /**
     * Get list of available datasets
     */
    getAvailableDatasets() {
        return this.datasets;
    }
    
    /**
     * Download a Wikipedia dataset with redirect handling
     */
    downloadDataset(datasetName, progressCallback = null) {
        return new Promise((resolve, reject) => {
            if (!this.datasets[datasetName]) {
                reject(new Error(`Unknown dataset: ${datasetName}`));
                return;
            }
            
            const dataset = this.datasets[datasetName];
            const filename = path.join(this.dataDir, `${datasetName}_wikipedia.xml.bz2`);
            
            console.log(`📥 Downloading ${dataset.name} from ${dataset.url}`);
            
            const downloadWithRedirects = (url, redirectCount = 0) => {
                if (redirectCount > 5) {
                    reject(new Error('Too many redirects'));
                    return;
                }
                
                const client = url.startsWith('https') ? https : http;
                
                const request = client.get(url, (response) => {
                    // Handle redirects
                    if (response.statusCode >= 300 && response.statusCode < 400 && response.headers.location) {
                        console.log(`Following redirect to ${response.headers.location}`);
                        downloadWithRedirects(response.headers.location, redirectCount + 1);
                        return;
                    }
                    
                    if (response.statusCode !== 200) {
                        reject(new Error(`HTTP ${response.statusCode}: ${response.statusMessage}`));
                        return;
                    }
                    
                    const totalSize = parseInt(response.headers['content-length'], 10) || 0;
                    let downloaded = 0;
                    let lastProgress = 0;
                    
                    const fileStream = fs.createWriteStream(filename);
                    
                    response.on('data', (chunk) => {
                        downloaded += chunk.length;
                        if (progressCallback && totalSize > 0) {
                            const progress = (downloaded / totalSize) * 100;
                            // Only report every 1% to reduce log spam
                            if (progress - lastProgress >= 1) {
                                progressCallback(progress, downloaded, totalSize);
                                lastProgress = progress;
                            }
                        }
                    });
                    
                    response.pipe(fileStream);
                    
                    fileStream.on('finish', () => {
                        fileStream.close();
                        console.log(`\n✅ Downloaded ${dataset.name} to ${filename}`);
                        resolve(filename);
                    });
                    
                    fileStream.on('error', (err) => {
                        fs.unlink(filename, () => {});
                        reject(err);
                    });
                });
                
                request.on('error', (err) => {
                    fs.unlink(filename, () => {});
                    reject(err);
                });
                
                // 10 minute timeout for large files
                request.setTimeout(600000, () => {
                    request.destroy();
                    reject(new Error('Download timeout'));
                });
            };
            
            downloadWithRedirects(dataset.url);
        });
    }
    
    /**
     * Extract and process Wikipedia XML dump into SQLite database
     * Memory-optimized for low-memory environments
     */
    extractAndProcess(compressedFile, dbPath, progressCallback = null) {
        return new Promise((resolve, reject) => {
            console.log(`📝 Processing ${compressedFile} into ${dbPath}`);
            
            // Initialize database with memory-optimized settings
            const db = new WikipediaDatabase(dbPath);
            db.initialize();
            
            // Force garbage collection if available
            if (global.gc) {
                global.gc();
            }
            
            // Create read stream with smaller buffer
            const readStream = fs.createReadStream(compressedFile, {
                highWaterMark: 64 * 1024 // 64KB chunks instead of default 64KB
            });
            
            const decompressStream = unbzip2();
            
            // Create SAX parser with strict mode
            const saxStream = sax.createStream(true, { 
                trim: true,
                normalize: true,
                lowercase: false,
                xmlns: false,
                position: false
            });
            
            // State variables - minimize object allocations
            let currentElement = null;
            let pageTitle = '';
            let pageText = '';
            let pageId = '';
            let inPage = false;
            let articlesProcessed = 0;
            let articlesSkipped = 0;
            
            // Batch processing for better performance
            const BATCH_SIZE = 100;
            let batch = [];
            
            saxStream.on('opentag', (node) => {
                const tagName = node.name;
                if (tagName === 'page') {
                    inPage = true;
                    pageTitle = '';
                    pageText = '';
                    pageId = '';
                } else if (inPage && (tagName === 'title' || tagName === 'text' || tagName === 'id')) {
                    currentElement = tagName;
                }
            });
            
            saxStream.on('text', (text) => {
                if (!currentElement || !inPage) return;
                
                if (currentElement === 'title') {
                    pageTitle += text;
                } else if (currentElement === 'text') {
                    // Limit text accumulation to prevent memory issues
                    if (pageText.length < 100000) {
                        pageText += text;
                    }
                } else if (currentElement === 'id' && !pageId) {
                    pageId = text;
                }
            });
            
            saxStream.on('cdata', (cdata) => {
                if (currentElement === 'text' && inPage && pageText.length < 100000) {
                    pageText += cdata;
                }
            });
            
            saxStream.on('closetag', (tagName) => {
                if (tagName === 'page' && inPage) {
                    inPage = false;
                    
                    // Check if valid article
                    if (isValidArticle(pageTitle, pageText)) {
                        // Process article
                        const cleanedContent = cleanWikipediaMarkup(pageText);
                        const summary = extractSummary(cleanedContent);
                        const categories = extractCategories(pageText);
                        const wordCount = cleanedContent.split(/\s+/).length;
                        
                        batch.push({
                            articleId: pageId,
                            title: pageTitle.trim(),
                            content: cleanedContent,
                            summary: summary,
                            categories: categories,
                            wordCount: wordCount
                        });
                        
                        articlesProcessed++;
                        
                        // Insert batch when full
                        if (batch.length >= BATCH_SIZE) {
                            db.insertBatch(batch);
                            batch = [];
                            
                            // Progress update
                            if (articlesProcessed % 1000 === 0) {
                                process.stdout.write(`\r📝 Processed ${articlesProcessed.toLocaleString()} articles (${articlesSkipped.toLocaleString()} skipped)`);
                                
                                // Periodic garbage collection hint
                                if (articlesProcessed % 10000 === 0 && global.gc) {
                                    global.gc();
                                }
                            }
                            
                            if (progressCallback) {
                                progressCallback(articlesProcessed);
                            }
                        }
                    } else {
                        articlesSkipped++;
                    }
                    
                    // Clear page data immediately
                    pageTitle = '';
                    pageText = '';
                    pageId = '';
                    currentElement = null;
                } else if (tagName === currentElement) {
                    currentElement = null;
                }
            });
            
            saxStream.on('error', (err) => {
                console.error('SAX parsing error:', err.message);
                // Continue processing despite errors
                saxStream._parser.error = null;
                saxStream._parser.resume();
            });
            
            saxStream.on('end', () => {
                // Insert remaining batch
                if (batch.length > 0) {
                    db.insertBatch(batch);
                    batch = [];
                }
                
                console.log(`\n📊 Creating full-text search index...`);
                console.log(`📚 Processed ${articlesProcessed.toLocaleString()} articles`);
                console.log(`⏭️  Skipped ${articlesSkipped.toLocaleString()} non-article pages`);
                
                db.createSearchIndex(() => {
                    const stats = db.getStatsSync();
                    console.log(`✅ Processing completed!`);
                    console.log(`📚 Total articles in database: ${stats.total_articles.toLocaleString()}`);
                    
                    db.close();
                    resolve(dbPath);
                });
            });
            
            // Pipe streams together
            readStream
                .pipe(decompressStream)
                .pipe(saxStream);
            
            readStream.on('error', (err) => {
                db.close();
                reject(err);
            });
            
            decompressStream.on('error', (err) => {
                db.close();
                reject(err);
            });
        });
    }
}

/**
 * Check if page is a valid article
 */
function isValidArticle(title, text) {
    if (!title || !text) return false;
    
    // Skip special pages
    if (title.startsWith('Category:') ||
        title.startsWith('Template:') ||
        title.startsWith('File:') ||
        title.startsWith('Wikipedia:') ||
        title.startsWith('Module:') ||
        title.startsWith('MediaWiki:') ||
        title.startsWith('Draft:') ||
        title.startsWith('Portal:') ||
        title.startsWith('Help:') ||
        title.startsWith('User:') ||
        title.startsWith('Talk:')) {
        return false;
    }
    
    // Skip redirects and stubs
    const textUpper = text.substring(0, 100).toUpperCase();
    if (textUpper.includes('#REDIRECT')) {
        return false;
    }
    
    // Skip very short articles
    if (text.trim().length < 200) {
        return false;
    }
    
    return true;
}

/**
 * WikipediaDatabase - SQLite database management with memory optimization
 */
class WikipediaDatabase {
    constructor(dbPath) {
        this.dbPath = dbPath;
        this.db = null;
    }
    
    /**
     * Initialize database schema with memory-optimized settings
     */
    initialize() {
        this.db = new sqlite3.Database(this.dbPath);
        
        // Memory-optimized PRAGMA settings
        this.db.serialize(() => {
            // Use smaller cache to reduce memory
            this.db.run('PRAGMA cache_size = 2000'); // ~2MB cache
            this.db.run('PRAGMA page_size = 4096');
            this.db.run('PRAGMA journal_mode = WAL');
            this.db.run('PRAGMA synchronous = NORMAL');
            this.db.run('PRAGMA temp_store = FILE'); // Use disk for temp storage
            this.db.run('PRAGMA mmap_size = 0'); // Disable memory mapping
            
            // Create tables
            this.db.run(`
                CREATE TABLE IF NOT EXISTS wikipedia_articles (
                    id INTEGER PRIMARY KEY,
                    article_id TEXT UNIQUE,
                    title TEXT NOT NULL,
                    content TEXT NOT NULL,
                    summary TEXT,
                    categories TEXT,
                    word_count INTEGER,
                    created_at DATETIME DEFAULT CURRENT_TIMESTAMP
                )
            `);
            
            this.db.run(`
                CREATE TABLE IF NOT EXISTS wikipedia_metadata (
                    key TEXT PRIMARY KEY,
                    value TEXT
                )
            `);
            
            this.db.run('CREATE INDEX IF NOT EXISTS idx_title ON wikipedia_articles(title)');
            this.db.run('CREATE INDEX IF NOT EXISTS idx_article_id ON wikipedia_articles(article_id)');
        });
    }
    
    /**
     * Insert a batch of articles efficiently
     */
    insertBatch(articles) {
        const stmt = this.db.prepare(`
            INSERT OR REPLACE INTO wikipedia_articles 
            (article_id, title, content, summary, categories, word_count)
            VALUES (?, ?, ?, ?, ?, ?)
        `);
        
        this.db.serialize(() => {
            this.db.run('BEGIN TRANSACTION');
            
            for (const article of articles) {
                stmt.run(
                    article.articleId,
                    article.title,
                    article.content,
                    article.summary,
                    article.categories,
                    article.wordCount
                );
            }
            
            this.db.run('COMMIT');
        });
        
        stmt.finalize();
    }
    
    /**
     * Create full-text search index
     */
    createSearchIndex(callback) {
        console.log('✅ Database ready (no FTS index - using direct queries)');
        if (callback) callback();
    }
    
    /**
     * Get database statistics synchronously
     */
    getStatsSync() {
        try {
            // Use synchronous prepare/bind/step for actual sync behavior
            const stmt = this.db.prepare('SELECT COUNT(*) as count FROM wikipedia_articles');
            let result = { total_articles: 0, total_words: 0 };
            
            if (stmt.step()) {
                const row = stmt.getAsObject();
                result.total_articles = row.count || 0;
            }
            
            stmt.free();
            return result;
        } catch (e) {
            console.error('Stats error:', e.message);
            return { total_articles: 0, total_words: 0 };
        }
    }
    
    /**
     * Close database connection
     */
    close() {
        if (this.db) {
            this.db.close();
            this.db = null;
        }
    }
}

/**
 * Download and process Wikipedia for server startup
 * @param {string} datasetName - Dataset to download (default: 'simple')
 * @param {string} dbPath - Path for SQLite database
 * @param {string} dataDir - Directory for temporary data
 * @returns {Promise<string>} Path to database
 */
async function downloadAndProcessWikipedia(datasetName = 'simple', dbPath = './wikipedia.db', dataDir = './wikipedia_data') {
    const downloader = new WikipediaDownloader(dataDir);
    
    // Progress callback for download
    const downloadProgress = (progress, downloaded, total) => {
        const downloadedMB = (downloaded / 1024 / 1024).toFixed(1);
        const totalMB = (total / 1024 / 1024).toFixed(1);
        process.stdout.write(`\r📥 Download progress: ${progress.toFixed(1)}% (${downloadedMB}MB / ${totalMB}MB)`);
    };
    
    // Progress callback for processing
    const processProgress = (articlesProcessed) => {
        // Progress is logged in the processor
    };
    
    try {
        // Step 1: Download
        console.log('📥 Step 1/2: Downloading Wikipedia dump...');
        const compressedFile = await downloader.downloadDataset(datasetName, downloadProgress);
        
        // Step 2: Process
        console.log('\n📝 Step 2/2: Processing Wikipedia articles...');
        console.log('⏱️  This will take several minutes...');
        const result = await downloader.extractAndProcess(compressedFile, dbPath, processProgress);
        
        // Clean up compressed file to save space
        console.log('🧹 Cleaning up temporary files...');
        try {
            fs.unlinkSync(compressedFile);
            // Also try to remove the data directory if empty
            fs.rmdirSync(dataDir);
        } catch (e) {
            // Ignore cleanup errors
        }
        
        return result;
    } catch (error) {
        console.error(`\n❌ Wikipedia download/processing failed: ${error.message}`);
        throw error;
    }
}

module.exports = {
    WikipediaDownloader,
    WikipediaDatabase,
    downloadAndProcessWikipedia,
    DATASETS,
    cleanWikipediaMarkup,
    extractSummary,
    extractCategories,
    isValidArticle
};
